actors/miner/src/deadline_state.rs

```diff
@@ -99,6 +103,127 @@ impl Deadlines {
         self.due[deadline_idx as usize] = store.put_cbor(deadline, Code::Blake2b256)?;
         Ok(())
     }
+
+    pub fn move_partitions<BS: Blockstore>(
+        &mut self,
+        store: &BS,
+        from_deadline: &mut Deadline,
+        to_deadline: &mut Deadline,
+        to_quant: QuantSpec,
+        partitions: &BitField,
+    ) -> anyhow::Result<()> {
+        let mut from_partitions = from_deadline.partitions_amt(store)?;
+        let mut to_partitions = to_deadline.partitions_amt(store)?;
+
+        // even though we're moving partitions intact, we still need to update from/to `Deadline` accordingly.
+
+        let first_to_partition_idx = to_partitions.count();
+        for (i, from_partition_idx) in partitions.iter().enumerate() {
+            let mut moving_partition = from_partitions
+                .get(from_partition_idx)?
+                .ok_or_else(|| actor_error!(not_found, "no partition {}", from_partition_idx))?
+                .clone();
+            if !moving_partition.faults.is_empty() || !moving_partition.unproven.is_empty() {
+                return Err(actor_error!(forbidden, "partition with faults or unproven sectors are not allowed to move, partition_idx {}", from_partition_idx))?;
+            }
+
+            let to_partition_idx = first_to_partition_idx + i as u64;
+
+            let from_expirations_epochs: Array<ExpirationSet, _> =
+                Array::load(&moving_partition.expirations_epochs, store)?;
+
+            let from_early_terminations: Array<BitField, _> =
+                Array::load(&moving_partition.early_terminated, store)?;
+
+            let mut to_expirations_epochs = Array::<ExpirationSet, BS>::new_with_bit_width(
+                store,
+                PARTITION_EXPIRATION_AMT_BITWIDTH,
+            );
+            let mut to_early_terminations = Array::<BitField, BS>::new_with_bit_width(
+                store,
+                PARTITION_EARLY_TERMINATION_ARRAY_AMT_BITWIDTH,
+            );
+
+            // ajust epoch info of expirations_epochs and early_terminated within `Partition`
+            from_expirations_epochs.for_each(|from_epoch, expire_set| {
+                to_expirations_epochs.set(
+                    to_quant.quantize_up(from_epoch as ChainEpoch) as u64,
+                    expire_set.clone(),
+                )?;
+                Ok(())
+            })?;
+            from_early_terminations.for_each(|from_epoch, bitfield| {
+                to_early_terminations
+                    .set(to_quant.quantize_up(from_epoch as ChainEpoch) as u64, bitfield.clone())?;
+                Ok(())
+            })?;
+            moving_partition.expirations_epochs = to_expirations_epochs.flush()?;
+            moving_partition.early_terminated = to_early_terminations.flush()?;
```

Comment:
Move this code into a method on partition state. It will be easier to test. Then you won't need to make PARTITION_EXPIRATION_AMT_BITWIDTH public either.

Reply:
Added `Partition::adjust_for_move` for that.